### PR TITLE
Feature/issue 15756 projection variables

### LINKED ISSUE #15756: Waypoint projection does not instantiate variables
### Describe your problem!

Using variables in waypoint projection (see #12600) does not create variables.

### How to reproduce?

1. Create a new waypoint
2. Set starting coordinates
3. Select "projection"
4. Enter variables in heading/distance
5. Save waypoint

### Actual result after these steps?

No variables were created in the variables page.

### Expected result after these steps?

Variables should be created in the variables page like it's done for the regular waypoint calculator. Additionally it would be nice to have the same quick access to variables as in the waypoint calculator.

### Reproducible

Yes

### c:geo Version

2024.05.11-NB-c926353

### System information

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by murggel**
## Description

This PR creates variables used in projection and shows them in the EditWaypointActivity.

I looked around in CoordinatesCalculateGlobalDialog and tried to used common logic.

## Related issues

fix #15756

## Additional context

grafik

### CODE REVIEW

**Comment by murggel** [comment]
I will fix the codacy-issues first

File: main/src/main/java/cgeo/geocaching/EditWaypointActivity.java [hunk 1]
```diff
@@ -391,11 +397,29 @@ public static ImmutableTriple<FormulaEditText, FormulaEditText, TextSpinner<Dist
         }
     }
 
+    private Set<String> getNeededVariablesForProjection() {
+        final Set<String> projectionVars = new HashSet<>();
+        final ProjectionType pt = projectionType.get();
+        if (pt == ProjectionType.BEARING) {
+            FormulaUtils.addNeededVariables(projectionVars, binding.projectionBearingDistance.getFormulaText());
```
Comment (eddiemuc) [inline main/src/main/java/cgeo/geocaching/EditWaypointActivity.java:404]: `FormulaUtils.addNeededVariables` will unnecessarily recompile the formula while it is already available in `FormulaEditText`. Suggestion: add a method `getNeededVars` to class `FormulaEditText` and use this method instead.

**Review by eddiemuc — APPROVED** [review]
lgtm. comment is not critical. Thanks for your work!

File: main/src/main/java/cgeo/geocaching/EditWaypointActivity.java (hunk 1 above)
Comment (murggel) [reply main/src/main/java/cgeo/geocaching/EditWaypointActivity.java]: done

File: main/src/main/java/cgeo/geocaching/EditWaypointActivity.java [hunk 2]
```diff
@@ -391,11 +396,28 @@ public static ImmutableTriple<FormulaEditText, FormulaEditText, TextSpinner<Dist
         }
     }
 
+    private Set<String> getNeededVariablesForProjection() {
+        final Set<String> projectionVars = new HashSet<>();
+
+        final ImmutableTriple<FormulaEditText, FormulaEditText, TextSpinner<DistanceUnit>> fields = getFieldsForProjectionType();
+        if (fields.left == null) {
```
Comment (eddiemuc) [inline main/src/main/java/cgeo/geocaching/EditWaypointActivity.java:403]: I assume the check has to be then other way around? (`... !=null`)
As is the code will produce an NPE in next line if `fields.left` is null

Comment (murggel) [reply main/src/main/java/cgeo/geocaching/EditWaypointActivity.java:403]: Ups

Comment (murggel) [reply main/src/main/java/cgeo/geocaching/EditWaypointActivity.java:403]: And the vars will not be added :-o

File: main/src/main/java/cgeo/geocaching/EditWaypointActivity.java [hunk 3]
```diff
@@ -391,11 +396,28 @@ public static ImmutableTriple<FormulaEditText, FormulaEditText, TextSpinner<Dist
...
+            fields.left.addNeededVariables(projectionVars);
+        }
+        if (fields.middle == null) {
```
Comment (murggel) [inline main/src/main/java/cgeo/geocaching/EditWaypointActivity.java:406]:
```suggestion
        if (fields.middle != null) {
```

File: main/src/main/java/cgeo/geocaching/EditWaypointActivity.java (hunk 2 above)
Comment (murggel) [inline main/src/main/java/cgeo/geocaching/EditWaypointActivity.java:403]:
```suggestion
        if (fields.left != null) {
```

**Comment by eddiemuc** [comment]
Lgtm, I will merge. Thanks for your work!

**Pull request merged**